aesm-client/src/imp/aesm_protobuf/mod.rs

```diff
@@ -107,4 +109,150 @@ impl AesmClient {
+                         pub_key_size : res.get_pub_key_id_size(),
+        })
+    }
+
+    ///
+    /// Returns information needed by an intel SGX application to get a quote of one of its enclaves.
+    /// Similar functionality to sgx_init_quote_ex in page 165 at https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Developer_Reference_Linux_2.9.1_Open_Source.pdf
+    ///
+    /// @param[in] att_key_id   - a byte array with hidden type 'sgx_att_key_id_ext_t' returned by AesmClient::select_att_key_id or AesmKeyIds::select_algorithm_id or similar.
+    ///
+    /// @returns size of the quote buffer
+    pub fn get_quote_size_ex(&self, att_key_id: Vec<u8>) -> Result<Option<u32>> {
+        let mut req = Request_GetQuoteSizeExRequest::new();
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+        req.set_att_key_id(att_key_id);
+
+        let res = self.transact(req)?;
+
+        let quote_size : Option<u32> = if res.has_quote_size() { Some(res.get_quote_size()) } else { None };
+        Ok(quote_size)
+    }
+
+    ///
+    /// takes the application enclave REPORT and generates a QUOTE.
+    /// Similar functionality to sgx_get_quote_ex in page 173 at https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Developer_Reference_Linux_2.9.1_Open_Source.pdf
+    ///
+    /// @param[in] report - Report of the enclave for that the quote is being calculated.
+    /// @param[in] att_key_id - Selected attestation key ID returned by sgx_select_att_key_id (or one selected from output of get_supported_att_key_ids). 
+    /// @param[in] nonce      - information required to generate a REPORT that can be verified by the application enclave.
+    /// @param[in] buf_size  Size of the reply quote byte array returned via 'get_quote_size_ex'.
+    ///                      Must be greater then SE_QUOTE_LENGTH_WITHOUT_SIG + sign_size where sign_size is variable (linux-sgx/psw/ae/qe/quoting_enclave.cpp:1069)
+    ///                      Otherwise will return parameter error.
+    ///
+    pub fn get_quote_ex(&self, report: Vec<u8>, quote_info : QuoteInfoEx, nonce: &[u8; 16], buf_size: u32) -> Result<QuoteResult> {
+        let mut req = Request_GetQuoteExRequest::new();
+        
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        req.set_att_key_id(quote_info.att_key_id);
+        req.set_report(report.clone());
+
+        // qe_report_info is of type sgx_ql_qe_report_info_t
+        // - nonce: sgx_quote_nonce_t
+        // - app_enclave_target_info: sgx_target_info_t
+        // - qe_report: sgx_report_t
+
+        let mut qe_report_info : Vec<u8> = Vec::new();
+
+        qe_report_info.extend(nonce.to_vec().clone());
```

Comment:
`.clone()` is redundant when calling `.to_vec()`. But even `to_vec` is unnecessary here.

Reply:
thanks!